Infer schema from gremlin/sparql query results for Neptune qpt

VenkatasivareddyTR opened the pull request:
*Issue #, if available:*

*Description of changes:*
Currently we are generating schema from aws glue for Neptune qpt as well
[Neptune_QPT_Tests.xlsx](https://github.com/user-attachments/files/18505470/Neptune_QPT_Tests.xlsx)
, this change will get schema from gremlin/sparql query results for Neptune qpt.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

chngpe commented on athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/NeptuneMetadataHandler.java:
```diff
@@ -296,7 +291,7 @@ public GetTableResponse doGetQueryPassthroughSchema(BlockAllocator allocator, Ge
                 GraphTraversalSource graphTraversalSource = neptuneConnection.getTraversalSource(client);
                 String gremlinQuery = qptArguments.get(NeptuneQueryPassthrough.QUERY);
-                gremlinQuery = gremlinQuery.concat(".limit(1)");
+                gremlinQuery = gremlinQuery.concat(".limit(10)");
```
Is it possible that the first 10 row data is different than the rest of rows?

Example:
Can first 10 rows contains a different result set which has no column field call `resource_type` and 11~20 rows has that? If yes, the schema inferred will drop the column? (and vice-versa, if no column in first 10 rows and data coming from 11~20, when reading data will there be an issue?)

It is okay, if not supported but i want to document the behavior as documentation

Thanks team

VenkatasivareddyTR replied:
Yes, we can update documentation about this limitation.

We determine the schema by examining the first 10 records. If the data exhibits schema variations in further rows, then we may not accurately capture the complete schema. It seems similar approach is followed in DynamoDB connector as well.

AbdulR3hman commented on athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/rdf/NeptuneSparqlConnection.java:
```diff
@@ -96,61 +96,66 @@ public boolean hasNext()
-                else if (type.equals(XMLSchema.LONG)) {
-                    oval = lit.longValue();
-                    logger.debug("Long " + varName + "=" + oval);
+            if (val != null) {
```
are we saying here that if value is null; we are still adding it to `ret` map?

AbdulR3hman replied:
alternatively; if we want know that if `value == null` then we return as is; why not check for that and `continue;` 

that way the next code is not indent and easier to read.

VenkatasivareddyTR replied:
> are we saying here that if value is null; we are still adding it to `ret` map?

Yes, to maintain consistency with SPARQL query results, which include column names with empty values, null values are also added to the 'ret' map.

VenkatasivareddyTR replied:
> alternatively; if we want know that if `value == null` then we return as is; why not check for that and `continue;`
> 
> that way the next code is not indent and easier to read.

If we use `value == null` then it's not consistent with SPARQL query results.
![image](https://github.com/user-attachments/assets/78117d16-73cb-44d8-a531-bc72c80d77d2)

SPARQL query results:
![image](https://github.com/user-attachments/assets/b81c6311-eb9c-495d-8753-69aa8362e3be)

With existing PR code changes:
![image](https://github.com/user-attachments/assets/03a878c9-47e8-4941-8b84-744153f89500)

AbdulR3hman replied:
understood; sounds good.

AbdulR3hman commented on athena-neptune/src/test/java/com/amazonaws/athena/connectors/neptune/NeptuneSchemaUtilsTest.java:
```diff
@@ -0,0 +1,64 @@
+public class NeptuneSchemaUtilsTest {
+    private static final Logger logger = LoggerFactory.getLogger(NeptuneSchemaUtilsTest.class);
+
+    private final Map<String, Object> objectMap = ImmutableMap.of(
```
can we also simulate a null as a test

VenkatasivareddyTR replied:
Hi @AbdulR3hman, If I understand correctly, a null value needs to be added to the objectMap? If so, I have added a null value to objectMap. 
Alternatively, if you expect the objectMap itself to be null, that is not applicable here, as it is a NotNull parameter.

VenkatasivareddyTR commented on athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/qpt/NeptuneQueryPassthrough.java:
```diff
@@ -56,7 +57,7 @@ public String getFunctionName()
     public List<String> getFunctionArguments()
     {
-        return Arrays.asList(DATABASE, COLLECTION, QUERY);
+        return Arrays.asList(DATABASE, COLLECTION, QUERY, COMPONENT_TYPE);
```
Thank you @AbdulR3hman for reviewing, I have added code to support multiple signatures for neptune qpt.

VenkatasivareddyTR commented on athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/NeptuneSchemaUtils.java:
```diff
@@ -0,0 +1,101 @@
+        return schema;
+    }
+
+    private static void buildSchema(String columnName, Object columnValue, SchemaBuilder schemaBuilder)
```
Hi @AbdulR3hman Yes, that's correct. 

For Gremlin queries, we are currently limiting support to valueMap() traversals. These traversals return Object types, which our connection logic explicitly converts into a Map representation.

For SPARQL queries, the direct response from Neptune is already structured as a Map.

Therefore, both Gremlin (after conversion) and SPARQL results converge to a Map type. This enables us to use a single, unified function for building the schema from these results.

fal-bharadwaj commented on athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/qpt/NeptuneGremlinQueryPassthrough.java:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    @Override
+    public void customConnectorVerifications(Map<String, String> engineQptArguments)
```
I think we should add more validations based on the arguments like null checks and gremlin syntax validations?

fal-bharadwaj commented on athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/rdf/RDFHandler.java:
```diff
@@ -174,11 +174,9 @@ else if (queryMode.equals(Constants.QUERY_MODE_CLASS)) {
 
         // get results
-        String strim = recordsRequest.getSchema().getCustomMetadata().get(Constants.SCHEMA_STRIP_URI);
-        boolean trimURI = strim == null ? false : Boolean.parseBoolean(strim);
```
Why did we delete this? We no longer need to trim the result?

fal-bharadwaj replied:
Not sure I understand, looks like we did use it in line 181 below and then removed it from being used in neptuneConnection.next()?

VenkatasivareddyTR replied:
In 181 we are passing to neptuneConnection.next() function, however we never used inside next() function. Inside next() function we are using class level variable. Please see below screenshot for your reference.

![image](https://github.com/user-attachments/assets/f2793653-53c5-4d77-97e5-2f992bb14a34)

fal-bharadwaj reviewed (changes requested):
Please take a look at the comments, thanks.

VenkatasivareddyTR commented on athena-neptune/src/main/java/com/amazonaws/athena/connectors/neptune/qpt/NeptuneGremlinQueryPassthrough.java:
```diff
@@ -0,0 +1,85 @@
+    public void customConnectorVerifications(Map<String, String> engineQptArguments)
+    {
+        // Verify no mixed operations (SPARQL and Gremlin in same request)
+        if (engineQptArguments.containsKey(QUERY)) {
```
Added validation for Unsupported gremlin query format. Null checks and empty checks are already handled in parent class.
We have removed Unsupported gremlin query format validation check from NeptuneMetadataHandler as we are handling here.

VenkatasivareddyTR commented:
> Please take a look at the comments, thanks. Some automated checks also failed, please review them as well.

Thank you @fal-bharadwaj for reviewing, we have addressed your review comments, please let us know if you have any other review comments.

fal-bharadwaj reviewed (approved):
looks good, thanks for addressing the comments!
